server/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -1695,8 +1697,29 @@ public AliasFilter buildAliasFilter(ClusterState state, String index, Set<String
     /**
      * Returns a new {@link QueryRewriteContext} with the given {@code now} provider
      */
-    public QueryRewriteContext getRewriteContext(LongSupplier nowInMillis) {
-        return new QueryRewriteContext(parserConfig, client, nowInMillis);
+    public QueryRewriteContext getRewriteContext(LongSupplier nowInMillis, IndicesRequest indicesRequest) {
+        // Skip unavailable indices
+        // TODO: Support metadata lookups for indices on remote clusters
+        IndicesOptions.Builder indicesOptionsBuilder = IndicesOptions.builder(indicesRequest.indicesOptions())
+            .concreteTargetOptions(new IndicesOptions.ConcreteTargetOptions(true));
+        Index[] indices = indexNameExpressionResolver.concreteIndices(
+            clusterService.state(),
+            indicesOptionsBuilder.build(),
+            true,
+            indicesRequest.indices()
+        );
+
+        Map<String, IndexMetadata> indexMetadataMap = new HashMap<>();
+        for (Index index : indices) {
+            IndexMetadata indexMetadata = clusterService.state().metadata().index(index);
```

Comment:
That's a big take for queries that don't require this. Exposing the list of concrete indices can also be quite big if we talk about data streams and index prefixes. 
In general I think we'll need to expose more primitive things that we can do on the coordinator rather than preemptively doing the work for everything since the new query will be used in a minority of rewrite. 
For instance we can expose the non-resolved indices, the index name expression resolver and the index metadata lookup?

Reply:
I updated this logic to use a `Supplier` to get the index metadata map lazily. This encapsulates all the dependencies for index metadata lookup, keeping interfaces clean and testing simple.

WDYT?